### ISSUE TITLE
chaining permissions logic / wrapResolve

### ISSUE THREAD

**Issue by medv** [graphql-compose/graphql-compose#112]
schema.js
```
schemaComposer
  .rootMutation()
  .addFields({
    user_create: UserTC
      .get('$createOne')
      .wrapResolve(isAboveRank('ADMIN')),
    user_update: UserTC
      .get('$updateById')
      .wrapResolve(isAboveRank('ADMIN')),
    user_remove: UserTC
      .get('$removeById')
      .wrapResolve(isAboveRank('ADMIN')),
  })
```

permissions.js
```
export const isAboveRank = (rank) => {
  const ranks = [
    'BANNED',
    'USER',
    'MODERATOR',
    'ADMIN',
  ]

  return next =>
    (params) => {
      const { context: { user } } = params

      let allowed = false

      if (user) {
        const currentRank = ranks.indexOf(user.rank)
        const requiredRank = ranks.indexOf(rank)

        if (currentRank >= requiredRank) allowed = true
      }

      return allowed
        ? next(params)
        : Error('Forbidden')
    }
}
```

What is the convention to handle more complex cases? For example, all my User CRUD is doable only by admins, however a user should be able to act on their own User object. I can put that easily into the above function like so

```
if (source && source._id && source._id.toString() === user._id.toString()) allowed = true
```
But this couples the logic. What about anything the user owns? In theory, I could always keep `owner_id` property on any object and check `source.owner_id` vs `context.user._id` like above, and naming this wrapper `isOwner`. I can see how I already have at least two or three layers of checks to be made for some resolvers. Eg some resolver might be ran if `isAuthenticated` -> (`isAdmin` OR `isMe`) or `isAuthenticated` -> (`isAdmin` OR `isOwner`). At the moment, of course my first wrapping function throws an error and the rest of the chain does not run (and as you can see, I'm coupling the `isAuthenticated` logic into `isAboveRank` it as well, by first checking if `context.user` is set). What about objects that are deeply nested? For example if there is a concept of a group, then within it a team and within it a todo list, then simply checking `owner_id` on the todo item is insufficient and we must trickle upward to see what sort of level of access you might have above.

In short, are there some common patterns that I am not aware of?

**Comment by nodkz** [comment]
I can suggest two patterns, both of them we are using in our internal app.

### 1. Combine permission logic in one resolver.
You should check all permissions before throwing an error.
```js
next => rp => {
  // show Data for owner or admin
  if (!isOwner(rp) && !isAdmin(rp)) {
    return null; // or Error
  }
  return next(rp);
}
``` 

### 2. Provide different endpoints for your admins and owners in mutations:
```js
const AdminMutationsTC = TypeComposer.create('AdminMutations');
AdminMutationsTC.addFields({
    user_create: UserTC
      .getResolver('createOne')
      .wrapResolve(isAboveRank('ADMIN')),
  // ...
});

const OwnerMutationsTC = TypeComposer.create('OwnerMutations');
OwnerMutationsTC.addFields({
    user_create: UserTC
      .getResolver('createOne')
      .wrapResolve(isOwner()),
  // ...
});

schemaComposer
  .rootMutation()
  .addFields({
    adminMutations: { type: AdminMutationsTC, resolve: () => ({}) },
    ownerMutations: { type: OwnerMutationsTC, resolve: () => ({}) }, 
  })
```

Moreover how I see you check admin permissions from `context`, so you may check access rights only once and do not wrap resolvers under `AdminMutationsTC`:
```js
schemaComposer
  .rootMutation()
  .addFields({
    adminMutations: { type: AdminMutationsTC, resolve: (_, __, context) => { 
      if (isAdmin(context.user)) return {};  // `{}` will call underlying field resolving
      throw new Error('You are not admin'); // or `return null;` it will not call underlying fields for resolving
    } },
  })
```

It's pitty that for `owner` cannot be so easy checked on top level. Cause required to load record from DB and checking ownership. So need to wrap all resolvers for checking owner permissions.

**Comment by medv** [comment]
Thanks for such quick response.

So in the case of nesting (group -> team -> item), and having to check permissions for every layer above the item (I may not be the item owner, but I might be a group moderator and therefore have access), is it correct that I have to fetch the whole tree and check on each level?

I guess I do not mind writing specific permissions resolvers for every case, is this a big performance blocker in general? For a single item that is fine, I may fetch 3 items from the database, but for a set of 100, that is 300 queries. Would DataLoader help with this at all, or am I overcomplicating things for myself and the actual answer is "you have to fetch this stuff if you want to use it"?

**Comment by nodkz** [comment]
TL;DR; No GraphQL, no DataLoader do not provide any kind of permission checks. There is only one place where you can do it - inside `resolve` method.

> is it correct that I have to fetch the whole tree and check on each level?

Yep, in your case, you are doing all right. 

> Would DataLoader help with this at all?

No. DataLoader only fetch data via bulk request. Then it provides obtained results granularly to every resolver, and there you need check permissions as you wish. You load all data, but `resolve` method decide to give it to the user or not.

**Comment by medv** [comment]
Ok thank you very much nodkz. You are actually insanely helpful as per the other issues here. Especially considering my question is more architectural as gql does not seem to be mapped so well publicly for advanced use cases yet.
